### PR TITLE
refactored mnemonic

### PR DESCRIPTION
**Pull request by jimmysong**
took out asserts
put words in a separate file
lazy initialization
functions to do word lookups instead of list/dict

### CODE REVIEW

File: buidl/mnemonic.py
Comment (mflaxman) [inline buidl/mnemonic.py:91]: Is a global needed? Can't this code block just run every time this file is loaded (outside the `_init_()`.

I don't have a lot of experience with globals and thought the best practice was to avoid them unless there was a strong reason. I could be wrong though.

File: buidl/mnemonic.py
Comment (mflaxman) [inline buidl/mnemonic.py:63]: Is there a performance impact here? I would think the performance bottleneck would be reading the `txt` file (which happens all at once), not iterating through the words.

Not a big deal.

**Review by mflaxman — APPROVED** [review]
I like the new error raising (not asserts), but am not sure I understand the purpose of the rest. it seems to be more complex now and I'd think this would be trickier for a new developer to intuit what's going on.

Another way to improve readability of the block of bip39_words would be to make a file called something like `bip39_wordlist.py` that had only one entry:
```
BIP39_WORDS = """
abandon
ability
able
...
zone
zoo
""".strip().split()
```

I don't feel strongly, feel free to merge it however you like. I appreciate the constant iteration!

**Comment by mflaxman** [comment]
Just noticed the branch name, are you adding SLIP39 support? Neat!

**Comment by jimmysong** [comment]
I'm working on it, yes. Need it for some backups as you so eloquently argued for =)

**Comment by jimmysong** [comment]
OK, refactored this and added basic SLIP39 support. Only supports one level, not two like the spec, but that seems like overkill to me. Multisig should be the first level and SLIP39 the second. I suppose a third level could be used there as well, but I'm not sure what sort of use case that would be.

BIP39 and SLIP39 are classes that inherit from each other to make the code DRY. I probably need to add more tests, but this is a good start.

File: buidl/mnemonic.py
Comment (mflaxman) [inline buidl/mnemonic.py:107]: Maybe `get_word(cls, index):` would be less ambiguous?

File: buidl/mnemonic.py
Comment (mflaxman) [inline buidl/mnemonic.py:99]: Maybe `get_index()` would be less ambiguous? On second thought, `get_index()` could imply that if there wasn't a matching word you return `None` instead of raising an exception.

I think the correct way to call this might just be `WordList.lookup[word]` (get rid of the helper method all together) instead of `WordList.index(word)`. This is why I find classes tricky, because the caller has to look deep under the hood to figure out how they work. The previous `BIP39_WORD_LIST[word]` seems strictly superior to the new class-based structure.

File: buidl/shamir.py
Comment (mflaxman) [inline buidl/shamir.py:183]: Delete commented out code

File: buidl/shamir.py
Comment (mflaxman) [inline buidl/shamir.py:203]: Delete commented out code

File: buidl/mnemonic.py
Comment (mflaxman) [inline buidl/mnemonic.py:88]: This should throw an exception/crash if for some reason it is unable to parse this file. I can't think of why that would happen, perhaps on some OS the file path works differently and the text file isn't there.

File: buidl/test/test_hd.py
Comment (mflaxman) [inline buidl/test/test_hd.py:408]: Repeated point from before, but using a method like this makes it non-obvious that it should throw an error if this is not a word, whereas a normal dictionary lookup doesn't have this ambiguity.

File: buidl/hd.py
Comment (mflaxman) [inline buidl/hd.py:380]: This comment predates this PR, but I noticed we are doing `for word in words:` 2x here, once above for the index and once down here to normalize the words. I think this loop should be done 1x in the same place.

File: buidl/mnemonic.py
Comment (mflaxman) [reply buidl/mnemonic.py:99]: Also, should this lookup be for `word.lower()` to be more forgiving?

I was thinking that would happen as part of the `normalize()` function, but in the current code-path this method is called first from `from_mnemonic()`.

**Review by mflaxman — APPROVED** [review]
Awesome to have slip39 support! I didn't audit the math behind that closely, just commenting on the structure/features.

Overall, I'm not in love with the new class-based structure for BIP39 (and SLIP39) word lists. I think the added ambiguity/complexity is a negative for application developers who don't have your depth of understanding. That said, I don't think it's the end of the world and defer to you. I love your pace of iteration and don't want to slow that down with bikeshedding.

Is the reason for not adding all the test vectors [example](https://github.com/trezor/python-shamir-mnemonic/blob/master/vectors.json#L24-L30) because of the 1-level? I agree the use-case for multi-level is very limited, but cross-compatibility with other implementations (i.e. Trezor) seems essential. I'd like to add this to multiwallet and have it work out of the box with Hermit for example (I have a PSBT branch with QR support that I'm working on). For the latter, buidl can only be used if it supports multi-level so that it can replace what's currently there.

As before, please merge when you see fit!

File: buidl/mnemonic.py
Comment (jimmysong) [reply buidl/mnemonic.py:88]: SLIP39 is 1024 words, unfortunately, though I can put expected number or something in there.

File: buidl/mnemonic.py
Comment (jimmysong) [reply buidl/mnemonic.py:99]: I was thinking `.index` is a method available to lists. Maybe I should subclass a data structure. Let me see if I can make it more usable that way.

File: buidl/mnemonic.py
Comment (jimmysong) [reply buidl/mnemonic.py:107]: I'll do `__getitem__` so it works more like a dict.

File: buidl/shamir.py
Comment (jimmysong) [reply buidl/shamir.py:183]: this is actually a comment to show that the formula is the same =)

File: buidl/shamir.py
Comment (jimmysong) [reply buidl/shamir.py:203]: same as above, though now that I look at it, maybe I can combine them.

File: buidl/test/test_hd.py
Comment (jimmysong) [reply buidl/test/test_hd.py:408]: yea, i'll use `__getitem__` so it works like a dictionary lookup.

**Comment by jimmysong** [comment]
The code will now be able to recover 2-layer shamir. It will not generate it, though. Useful for recovery.

Now every test case except the last one is in the unit test.

**Comment by jimmysong** [comment]
@mflaxman please take a look one more time to see if your concerns have been addressed.

File: buidl/test/test_shamir.py
Comment (mflaxman) [inline buidl/test/test_shamir.py:247]: Print statement

File: buidl/hd.py
Comment (mflaxman) [inline buidl/hd.py:359]: Outdated variable name in comment

**Review by mflaxman — APPROVED** [review]
Have still not audited the math, but code LGTM.

**Comment by mflaxman** [comment]
Wohoo! @jimmysong, want to add this to multiwallet.py?

**Pull request merged**
